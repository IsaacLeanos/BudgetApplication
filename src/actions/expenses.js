import uuid from 'uuid'
import database from '../firebase/firebase'


export const addExpense=(expense)=>{
    return{
        type:'ADD_EXPENSE',
        expense:expense
    }
}
export const startAddExpense=({description='',note='',amount=0,createdAt=0}={})=>{
    return(dispatch,getState)=>{
        const uid=getState().authen.uid
        const expense={description,note,amount,createdAt}
        return database.ref(`users/${uid}/expenses`).push(expense).then((ref)=>{
            dispatch(addExpense({
                id:ref.key,
                ...expense
            }))
        }).catch((e)=>{console.log(e)})
    }
}
export const startRemoveExpense=({id}={})=>{
    return(dispatch,getState)=>{
        const uid=getState().authen.uid
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense({id:id}))
        })
    }
}

export const removeExpense=({id}={})=>{
    return{
        type:'REMOVE_EXPENSE',
        id:id
    }
}
export const editExpense=(id,updates)=>{
    return{
        type:'EDIT_EXPENSE',
        id:id,
        updates
    }
}
// export const startEditExpense=(id,formData)=>{
//     const updates={amount,description,note,createdAt}
//     return(dispatch,getState)=>{
//         const uid=getState().authen.uid
//         return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(()=>{
//             dispatch(editExpense(id,updates))
//         })
//     }
// }
export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
      const uid = getState().authen.uid;
      return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
        dispatch(editExpense(id, updates));
      });
    };
  };

export const setExpenses=(expenses)=>{
    return{
        type:'SET_EXPENSES',
        expenses
    }
}
export const startSetExpenses=()=>{
    return(dispatch,getState)=>{
        const uid=getState().authen.uid
        return database.ref(`users/${uid}/expenses`).once('value').then((snapshot)=>{
            const expenses=[]
            snapshot.forEach((childSnapshot)=>{
                expenses.push({id:childSnapshot.key,...childSnapshot.val()})
            })
            dispatch(setExpenses(expenses))
        })
    }
}
