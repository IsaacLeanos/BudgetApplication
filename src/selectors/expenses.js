import moment from 'moment'


const selectExpenses=(expenses,{text,sortBy,startDate,endDate})=>{
    return expenses.filter((item)=>{
        let createdAtMoment=moment(item.createdAt)
        let startDateMatch=startDate?startDate.isSameOrBefore(createdAtMoment,'day'):true
        let endDateMatch=endDate?endDate.isSameOrAfter(createdAtMoment,'day'):true
        let textMatch=item.description.toLowerCase().includes(text.toLowerCase())
        return startDateMatch&&endDateMatch&&textMatch
    }).sort((a,b)=>{
        if(sortBy==='date'){
            return a.createdAt<b.createdAt ? -1:1
        }else if(sortBy==='amount'){
            return a.amount<b.amount ? 1:-1
        }
    })
}

export default selectExpenses