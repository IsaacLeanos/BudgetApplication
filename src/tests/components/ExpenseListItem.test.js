import React from 'react'
import{shallow}from'enzyme'
import ExpenseListItem from '../../components/ExpenseListItem'
import expenses from '../fixtures/expensesData'



test('render expense link and attributes',()=>{
    let wrapper=shallow(<ExpenseListItem {...expenses[0]}/>)
    expect(wrapper).toMatchSnapshot()
})

