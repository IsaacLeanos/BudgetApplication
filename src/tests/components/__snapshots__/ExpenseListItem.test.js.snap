// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render expense link and attributes 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/123"
>
  <div>
    <h3
      className="list-item__title"
    >
      bll
    </h3>
    <span
      className="list-item__sub-Title"
    >
      December 31st, 1969
    </span>
  </div>
  <h3
    className="list-item__data"
  >
    $500.00
  </h3>
</Link>
`;
